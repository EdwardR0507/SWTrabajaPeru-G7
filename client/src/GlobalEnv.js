module.exports = {
  //host: process.env.HOST || "http://localhost:4000",
  host: "https://ws-trabajaperu.herokuapp.com",
};
