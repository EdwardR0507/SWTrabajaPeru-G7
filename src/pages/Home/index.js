import React, { useState, useEffect } from "react";
import { useLocation } from "react-router";
import { withStyles } from "@material-ui/styles";
import { Container, Typography, IconButton } from "@material-ui/core/";
import NavigateBeforeIcon from "@material-ui/icons/NavigateBefore";
import NavigateNextIcon from "@material-ui/icons/NavigateNext";
import NavBar from "../../layouts/NavBar";
import ServiceCard from "../../components/Cards/ServiceCard";
import WorkerCard from "../../components/Cards/WorkerCard";
import theme from "../../themes/themes";
import { fetchData, fetchUserData } from "../../services/services";
const StyledContentContainer = withStyles({
  root: {
    marginTop: "40px",
    display: "flex",
    justifyContent: "space-between",
  },
})(Container);

const StyledCardContainer = withStyles({
  root: {
    margin: "0 10px",
    display: "flex",
  },
})(Container);

const StyledIconButton = withStyles({
  root: {
    height: "50px",
    margin: "auto 0",
    backgroundColor: theme.palette.secondary.main,
    color: "white",
    "&:hover": {
      backgroundColor: theme.palette.secondary.main,
    },
  },
})(IconButton);

const StyledH2 = withStyles({
  root: {
    fontSize: "36px",
  },
})(Typography);

export default function Home() {
  //Paginación
  const [pageServices, setPageServices] = useState(0);
  const [pageWorkers, setPageWorkers] = useState(0);

  const location = useLocation();
  const [services, setServices] = useState();
  // Servicios que se reenderizan por página
  const [workers, setWorkers] = useState();
  // Trabajadores que se renderizan por página
  const [user, setUser] = useState();
  const state = location.state;

  useEffect(() => {
    fetchUserData("GET", "service", "GET_HOME_SERVICES").then((res) => {
      console.log(res);
      setServices(res);
    });
  }, []);

  useEffect(() => {
    fetchUserData("GET", "user", "GET_HOME_USERS").then((res) => {
      console.log(res);
      setWorkers(res);
    });
  }, []);

  let token;

  useEffect(() => {
    console.log(localStorage.getItem("User_session"))
    if (localStorage.hasOwnProperty("User_session")) {
      token = localStorage.getItem("User_session")
      token = token.slice(1, -1)
      fetchData(token, "GET", "user-auth", "GET_MY_USER")
        .then((res) => {
          console.log(res);
          setUser(res);
        })
        .catch((err) => {
          console.log(err);
        });
    } else {
      return;
    }
  }, [state?.token]);

  const handleServiceBefore = () => {
    if (pageServices > 0) {
      setPageServices(pageServices - 1);
    }
  };

  const handleWorkerBefore = () => {
    if (pageWorkers > 0) {
      setPageWorkers(pageWorkers - 1);
    }
  };

  const handleServiceAfter = () => {
    if (pageServices < services.length - 4) {
      setPageServices(pageServices + 1);
    }
  };

  const handleWorkerAfter = () => {
    if (pageWorkers < workers.length - 4) {
      setPageWorkers(pageWorkers + 1);
    }
  };

  const conditionalNavBar = () => {
    return user ? <NavBar user={user} token={token} /> : <NavBar />;
  }

  return workers && services ? (
    <>
      {conditionalNavBar()}
      <Container role="home">
        <StyledContentContainer>
          <StyledH2 variant="h2">Servicios</StyledH2>
        </StyledContentContainer>
        <StyledCardContainer>
          {/*Extensión a cantidad de servicios aleatorios
                       con información desde la base de datos 
                       Iterar y por cada info generar un service card*/}
          <StyledIconButton onClick={handleServiceBefore}>
            <NavigateBeforeIcon />
          </StyledIconButton>
          {services
            ?.slice(pageServices, pageServices + 3)
            .map((service) =>
              user ? (
                <ServiceCard
                  key={`${service.us_id}-${service.cat_id}`}
                  service={service}
                  token={token}
                />
              ) : (
                <ServiceCard
                  key={`${service.us_id}-${service.cat_id}`}
                  service={service}
                />
              )
            )}
          <StyledIconButton onClick={handleServiceAfter}>
            <NavigateNextIcon />
          </StyledIconButton>
        </StyledCardContainer>
        <StyledContentContainer>
          <StyledH2 variant="h2">Trabajadores</StyledH2>
        </StyledContentContainer>
        <StyledCardContainer>
          <StyledIconButton onClick={handleWorkerBefore}>
            <NavigateBeforeIcon />
          </StyledIconButton>
          {workers?.slice(pageWorkers, pageWorkers + 3).map((worker) => (
            <WorkerCard worker={worker} />
          ))}
          <StyledIconButton onClick={handleWorkerAfter}>
            <NavigateNextIcon />
          </StyledIconButton>
        </StyledCardContainer>
      </Container>
    </>
  ) : (
    <div>Cargando...</div>
  );
}
