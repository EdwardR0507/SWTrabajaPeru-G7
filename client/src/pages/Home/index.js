import React, { useState, useEffect } from "react";
import { useLocation } from "react-router";
import axios from "axios";
import GlobalEnv from "../../GlobalEnv";
import { withStyles } from "@material-ui/styles";
import { Container, Link, Typography, IconButton } from "@material-ui/core/";
import NavigateBeforeIcon from "@material-ui/icons/NavigateBefore";
import NavigateNextIcon from "@material-ui/icons/NavigateNext";
import NavBar from "../../layouts/NavBar";
import ServiceCard from "../../components/Cards/ServiceCard";
import WorkerCard from "../../components/Cards/WorkerCard";
import theme from "../../themes/themes";

const StyledContentContainer = withStyles({
  root: {
    marginTop: "40px",
    display: "flex",
    justifyContent: "space-between",
  },
})(Container);

const StyledCardContainer = withStyles({
  root: {
    margin: "0 10px",
    display: "flex",
  },
})(Container);

const StyledIconButton = withStyles({
  root: {
    height: "50px",
    margin: "auto 0",
    backgroundColor: theme.palette.secondary.main,
    color: "white",
    "&:hover": {
      backgroundColor: theme.palette.secondary.main,
    },
  },
})(IconButton);

const StyledH2 = withStyles({
  root: {
    fontSize: "36px",
  },
})(Typography);

const StyledLink = withStyles({
  root: {
    color: theme.palette.primary.main,
    marginRight: "40px",
    marginTop: "12px",
    letterSpacing: "0.18px",
  },
})(Link);

export default function Home() {
  const location = useLocation();
  const [services, setServices] = useState();
  const [workers, setWorkers] = useState();
  const state = location.state;

  useEffect(() => {
    //Cambiar post por get cuando se arregle
    axios
      .post(`${GlobalEnv.host}/service`, {
        command: "GET_HOME_SERVICES",
      })
      .then((res) => {
        console.log(res)
        setServices(res.data);
      });
  }, []);

  useEffect(() => {
    //Cambiar post por get cuando se arregle
    axios
      .post(`${GlobalEnv.host}/user`, {
        command: "GET_USERS",
      })
      .then((res) => {
        setWorkers(res.data);
      });
  }, []);

  return workers && services ? (
    <>
      {state ? <NavBar user={state.user} /> : <NavBar />}
      <Container>
        <StyledContentContainer>
          <StyledH2 variant="h2">Servicios</StyledH2>
          <StyledLink href="#">
            <Typography variant="h5">VER TODO</Typography>
          </StyledLink>
        </StyledContentContainer>
        <StyledCardContainer>
          {/*Extensión a cantidad de servicios aleatorios
                       con información desde la base de datos 
                       Iterar y por cada info generar un service card*/}
          <StyledIconButton>
            <NavigateBeforeIcon />
          </StyledIconButton>
          {services?.map((service) => (
            <ServiceCard service={service} />
          ))}
          <StyledIconButton>
            <NavigateNextIcon />
          </StyledIconButton>
        </StyledCardContainer>
        <StyledContentContainer>
          <StyledH2 variant="h2">Trabajadores</StyledH2>
          <StyledLink href="#">
            <Typography variant="h5">VER TODO</Typography>
          </StyledLink>
        </StyledContentContainer>
        <StyledCardContainer>
          <StyledIconButton>
            <NavigateBeforeIcon />
          </StyledIconButton>
          {workers?.map((worker) => (
            <WorkerCard worker={worker} />
          ))}
          <StyledIconButton>
            <NavigateNextIcon />
          </StyledIconButton>
        </StyledCardContainer>
      </Container>
    </>
  ) : (
    <div>Cargando...</div>
  );
}
