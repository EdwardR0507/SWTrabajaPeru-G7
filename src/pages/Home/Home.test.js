import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import {rest} from 'msw'
import {setupServer} from 'msw/node'
import { shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { render, cleanup } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { Router, Route } from 'react-router-dom'
import GlobalEnv from '../../GlobalEnv';
import Home from './index';  

describe("<Home />", () => {
    const serverUser = setupServer(
        rest.post(`${GlobalEnv.host}/user`, (req, res, ctx) => {
          return res(ctx.json({greeting: 'hello there'}))
        }),
      )
    const serverService = setupServer(
        rest.post(`${GlobalEnv.host}/service`, (req, res, ctx) => {
          return res(ctx.json({greeting: 'hello there'}))
        }),
      )

    beforeAll(() => {serverUser.listen(); serverService.listen()})
    afterEach(() => {serverUser.resetHandlers(); serverService.resetHandlers()})
    afterAll(() => {serverUser.close(); serverService.close()})

    it('render "Home"', () => {
        const history = createMemoryHistory()
        const wrapper = shallow(<Router history={history}>
            <Home /> 
        </Router>);
        expect(wrapper).toMatchSnapshot()
      });
    it("get User", () => {
        const history = createMemoryHistory();
        serverUser.use(
            rest.post(`${GlobalEnv.host}/user`, (req, res, ctx) => {
              return res(ctx.status(500))
            }),
          )
        render(<Router history={history}>
            <Home /> 
        </Router>);
        expect(screen.getByRole('home')).toHaveTextContent('Cargando...')
    })
    it("get Service", () => {
        const history = createMemoryHistory();
        serverService.use(
            rest.post(`${GlobalEnv.host}/service`, (req, res, ctx) => {
              return res(ctx.status(500))
            }),
          )
        const { getByRole } = render(<Router history={history}>
            <Home /> 
        </Router>);
        expect(screen.getByRole('home')).toHaveTextContent('Cargando...')
    })
})