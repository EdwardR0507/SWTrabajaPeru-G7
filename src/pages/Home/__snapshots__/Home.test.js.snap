// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> render "Home" 1`] = `ShallowWrapper {}`;
